test/client/tasks_test.rb

```diff
@@ -0,0 +1,179 @@
+require_relative "../test_helper"
+
+describe Elastomer::Client::Tasks do
+  before do
+    @tasks = $client.tasks
+  end
+
+  after do
+  end
+
+  it "list all in-flight tasks" do
+    h = @tasks.get
+    assert h["nodes"].keys.size > 0
+
+    total_tasks = h["nodes"].map { |k, v| v["tasks"].keys.count }.sum
+    assert total_tasks > 0
+  end
+
+  it "groups by parent->child relationships when get-all tasks API is grouped by 'parents'" do
+    unless $client.version_support.es_version_5_x?
+      skip "Tasks API is not supported in ES version #{$client.version}"
+    end
+
+    h = @tasks.get :group_by => "parents"
+    parent_id = h["tasks"].select { |k, v| v.key?("children") }.keys.first
+    childs_parent_ref = h.dig("tasks", parent_id, "children").first["parent_task_id"]
+    assert_equal parent_id, childs_parent_ref
+  end
+
+  it "raises exception when get_by_id is called without required task & node IDs" do
+    assert_raises(ArgumentError) do
+      @tasks.get_by_id
+    end
+  end
+
+  it "raises exception when get_by_id is called w/invalid task ID is supplied" do
+    node_id = @tasks.get["nodes"].map { |k, v| k }.first
+    assert_raises(ArgumentError) do
+      @tasks.get_by_id node_id, "task_id_should_be_integer"
+    end
+  end
+
+  it "raises exception when get_by_id is called w/invalid node ID is supplied" do
+    assert_raises(ArgumentError) do
+      @tasks.get_by_id nil, 42
+    end
+  end
+
+
+  it "successfully waits for task to complete when wait_for_completion and timeout flags are set" do
+    # do some busy work in bkg thread to keep some tasks running longer
+    Thread.new do
+      begin
+        name = "elastomer-tasks-tests".freeze
+        index = $client.index(name)
+        wait_for_index(name)
```

Comment:
Two things here for this test case:

1.  The thread here should create its own connection to the Elasticsearch cluster and use that in the test. The code outside the thread will also be sending requests which might stomp on the requests coming from inside the thread.

2. You need to `create()` the index first before calling the `wait_for_index` method.

```ruby
client = Elastomer::Client.new \
  host:         $client.host,
  port:         $client.port,
  read_timeout: $client.read_timeout,
  open_timeout: $client.open_timeout

index = client.index(name)
index.create({})
wait_for_index(name)
```